### DISCUSSION TITLE
Help with getting SMC to work on my example?

### DISCUSSION THREAD

**Discussion by sdwfrost** [probcomp/Gen.jl#537] [Q&A] [answered]
Hi everyone, thanks for Gen.jl! I've been having fun with it. I've been trying to get my head around the SMC example, and applying it to my problem. Can anyone give any pointers? Code using Metropolis Hastings is below; you can also find a hand-coded particle filter [here](https://github.com/epirecipes/sir-julia/blob/master/markdown/markov_pfilter/markov_pfilter.md) for reference.

```Julia
using OrdinaryDiffEq
using SciMLSensitivity
using Distributions
using Gen
using GenDistributions

# Define the Gen model
@gen function bayes_sir_markov(l::Int=40,
                               N::Int=1000,
                               c::Float64=10.0,
                               γ::Float64=0.25,
                               δt::Float64=1.0)
    i₀ ~ uniform_discrete(1, 100)
    β ~ uniform(0.01, 0.1)
    S = N - i₀
    I = i₀
    for i in 1:l
        ifrac = 1-exp(-β*c*I/N*δt)
        rfrac = 1-exp(-γ*δt)
        infection = {(:y, i)} ~ binom(S,ifrac)
        recovery = {(:z, i)} ~ binom(I,rfrac)
        S = S-infection
        I = I+infection-recovery
    end
    nothing
end

# Set the true parameters and solve the model
p = Gen.choicemap()
p[:β] = 0.05
p[:i₀] = 10
l = 40
(sol, _) = Gen.generate(bayes_sir_markov, (l,), p)

# Generate observations
Y = [sol[(:y, i)] for i=1:l]
observations = Gen.choicemap()
for (i, y) in enumerate(Y)
    observations[(:y, i)] = y
end

## Metropolis-Hastings

const truncated_normal = DistributionsBacked((mu,std,lb,ub) -> Distributions.Truncated(Normal(mu, std), lb, ub),
                                             (true,true,false,false),
                                             true,
                                             Float64)


@gen function sir_proposal(current_trace)
    β ~ truncated_normal(current_trace[:β], 0.01, 0.0, Inf)
    i₀ ~ uniform_discrete(current_trace[:i₀]-1, current_trace[:i₀]+1)
end

n_iter = 110000
burnin =  10000
thin = 10
β_mh = Vector{Real}(undef, (n_iter - burnin) ÷ thin)
i₀_mh = Vector{Int}(undef, (n_iter - burnin) ÷ thin)

(tr,) = Gen.generate(bayes_sir_markov, (l,), merge(observations, p))
for i in 1:n_iter
    (tr, _) = Gen.mh(tr, sir_proposal, ())
    if i <= burnin
        continue
    end
    if i % thin == 0
        β_mh[(i - burnin) ÷ thin] = tr[:β]
        i₀_mh[(i - burnin) ÷ thin] = tr[:i₀]
    end
end

# Posterior mean estimates
mean(β_mh) # cf 0.05
mean(i₀_mh) # cf 10
```

**Comment by ztangent** [comment]
Do you already have a partial implementation of SMC for this model?

In addition to what's in the docs and [the Gen tutorial on particle filtering](https://www.gen.dev/tutorials/particle-filtering/tutorial), you might be interested in using this SMC extension package for Gen called [GenParticleFilters.jl](https://github.com/probcomp/GenParticleFilters.jl) -- there's an example in the README for that package that is pretty similar to your hand-coded particle filter.

**Comment by sdwfrost** [comment]
Thanks for the tips @ztangent! I guess my main question is that for the model implementation above, my model doesn't return anything. In the SMC case, what (if anything) do I have to return?

**Reply by ztangent** [reply]
You don't have to return anything if you don't care about any particular return value! It's common to return the entire vector of latent variables and/or observations that were generated over time, but doing that is not necessary for the particle filter to perform inference.

**Reply by sdwfrost** [reply]
Thanks - I was getting confused, as I didn't know how the code 'knows' the timestep; in the tutorial example, the argument is just the total number of time steps, whereas I tend to think of SMC running on a timestep by timestep basis...

**Reply by ztangent** [reply]
That makes sense! Perhaps you've already figured this out, but the way it works is that at each timestep `t` the SMC algorithm increments the argument to the model (i.e. the total number of timesteps) from `t-1` to `t`, rerunning the entire model up to timestep `t`, but keeping fixed the latent variables that were sampled for all previous timesteps.

If you rewrite your model [using the Unfold combinator](https://www.gen.dev/tutorials/particle-filtering/tutorial#unfold), then Gen will also take advantage of [incremental computation](https://en.wikipedia.org/wiki/Incremental_computing) and avoid rerunning the model for previous timesteps (`1` to `t-1`). Instead, it will "extend" the model so that it runs only the most recent step `t`, ensuring that the particle filter has a runtime linear in the number of steps.

**Reply by sdwfrost** [reply]
Thanks - knowing the internal workings really helps. Another quick thing that isn't immediately apparent (at least to me) - in my MCMC example, I trace the tuples `(:y, i)` and `(:z,i)` (where an array of `y` corresponds to my observations. It seems that in the SMC examples, you trace dictionaries. To port my example to allow the internal constraints to work, do I need to write the model in terms of tracing `i => y` (or similar).

**Reply by ztangent** [reply]
You're still allowed to use tuples as addresses for your observations, as long as you're not using the Unfold combinator!

But if you use the Unfold combinator, it assumes a certain address structure. For example, if you write a time series model like this:

```julia
@gen function step(t::Int, prev_x::Real)
     x ~ normal(prev_x, 1)
     y ~ normal(x, 1)
end

unfold_step = Unfold(step)

@gen function model(T::Int)
    x_init ~ normal(0, 1)
    y_init ~ normal(x_init, 1)
    steps ~ unfold_step(T, x_init)
end
```

Then the address for the `y` observation at timestep `t` in `model` will be `:steps => t => :y`.

**Reply by sdwfrost** [reply]
Thanks @ztangent, for this example, it's cleaner not to use the Unfold combinator (though I'll think about it some more). I have something running now (the trick to do so was passing a vector of `ChoiceMap`s it seemed), although the variance of the parameters is way bigger than with MCMC, despite a respectable effective sample size - the stdev of `:i₀` is 27.7 vs. 1.5. Am I doing this correctly?

```Julia
using OrdinaryDiffEq
using SciMLSensitivity
using Distributions
using Random
using Gen
using GenDistributions
using GenParticleFilters

Random.seed!(1234)

# Define the Gen model
@gen function bayes_sir_markov(l::Int=40,
                               N::Int=1000,
                               c::Float64=10.0,
                               γ::Float64=0.25,
                               δt::Float64=1.0)
    i₀ ~ uniform_discrete(1, 100)
    β ~ uniform(0.01, 0.1)
    S = N - i₀
    I = i₀
    for i in 1:l
        ifrac = 1-exp(-β*c*I/N*δt)
        rfrac = 1-exp(-γ*δt)
        infection = {(:y, i)} ~ binom(S,ifrac)
        recovery = {(:z, i)} ~ binom(I,rfrac)
        S = S-infection
        I = I+infection-recovery
    end
    nothing
end

# Set the true parameters and solve the model
p = Gen.choicemap()
p[:β] = 0.05
p[:i₀] = 10
l = 40
(sol, _) = Gen.generate(bayes_sir_markov, (l,), p)

# Generate observations
Y = [sol[(:y, i)] for i=1:l]
observations = Gen.choicemap()
for (i, y) in enumerate(Y)
    observations[(:y, i)] = y
end

## Metropolis-Hastings

const truncated_normal = DistributionsBacked((mu,std,lb,ub) -> Distributions.Truncated(Normal(mu, std), lb, ub),
                                             (true,true,false,false),
                                             true,
                                             Float64)


@gen function sir_proposal(current_trace)
    β ~ truncated_normal(current_trace[:β], 0.01, 0.0, Inf)
    i₀ ~ uniform_discrete(current_trace[:i₀]-1, current_trace[:i₀]+1)
end

n_iter = 110000
burnin =  10000
thin = 10
β_mh = Vector{Real}(undef, (n_iter - burnin) ÷ thin)
i₀_mh = Vector{Int}(undef, (n_iter - burnin) ÷ thin)

(tr,) = Gen.generate(bayes_sir_markov, (l,), merge(observations, p))
for i in 1:n_iter
    (tr, _) = Gen.mh(tr, sir_proposal, ())
    if i <= burnin
        continue
    end
    if i % thin == 0
        β_mh[(i - burnin) ÷ thin] = tr[:β]
        i₀_mh[(i - burnin) ÷ thin] = tr[:i₀]
    end
end

# Posterior mean estimates
mean(β_mh), sqrt(var(β_mh)) # cf 0.05
mean(i₀_mh), sqrt(var(i₀_mh)) # cf 10

# Sequential Monte Carlo

function particle_filter(observations, n_particles, ess_thresh=0.5)
    # Initialize particle filter with first observation
    n_obs = length(observations)
    obs_choices = [choicemap((:y, t) => observations[t]) for t=1:n_obs]
    state = pf_initialize(bayes_sir_markov, (1,), obs_choices[1], n_particles)
    # Iterate across timesteps
    for t=2:n_obs
        # Resample if the effective sample size is too low
        if effective_sample_size(state) < ess_thresh * n_particles
            # Perform residual resampling, pruning low-weight particles
            pf_resample!(state, :residual)
        end
        # Update filter state with new observation at timestep t
        pf_update!(state, (t,), (UnknownChange(),), obs_choices[t])
    end
    return state
end

state = particle_filter(Y, 10000);
effective_sample_size(state)
mean(state, :β), sqrt(var(state, :β))
mean(state, :i₀), sqrt(var(state, :i₀))
```

**Reply by ztangent** [reply]
Hi, the code looks right to me! The one thing that I would double-check is whether `pf_update!` is handling default arguments correctly. `pf_update!` calls [`Gen.update`](https://www.gen.dev/docs/stable/ref/gfi/#Gen.update) under the hood, which has a particular way of handling default arguments. To be safe, I would explicitly specify all of the arguments to the model when calling `pf_initialize` and `pf_update!`, e.g.:

```julia
fixed_args = (1000, 10.0, 0.25, 1.0)
new_args = (t, fixed_args...)
argdiffs = (UnknownChange(), map(x -> NoChange(), new_args)...)
pf_update!(state, new_args, argdiffs, obs_choices[t])
```

Apart from that, there are a couple of things that could be going on that explains the difference in variance:
1. The MCMC chain might not be mixing well, and the SMC estimate might be closer to the true posterior.
2. The SMC algorithm might have collapsed on multiple bad hypotheses that are far apart, leading to high variance (though this is less likely because when particle filters collapse it's usually on a single particle).
3. There's some bug in the code that I'm not noticing.

One light-weight check to see whether Issue 1 is happening might be to run multiple MCMC chains and calculate the variance across all the chains. If it's much larger then single-chain MCMC, then it's likely that MCMC isn't mixing well with the current proposal.

A more involved check -- that a lot of us in our lab have been adopting as part of the model/inference debugging process -- is to first try and compute either the exact posterior, or a deterministic grid-based approximation to that posterior, just so that there's a gold standard to compare against. `GenParticleFilters` can actually help with this process, because it supports [stratified particle initialization](https://probcomp.github.io/GenParticleFilters.jl/dev/initialize/#Stratified-initialization) and [stratified particle updating](https://probcomp.github.io/GenParticleFilters.jl/dev/update/#Updating-with-stratified-sampling). Happy to explain more about how to use these features if there's still a big difference in variance.

**Reply by sdwfrost** [reply]
Thanks - I tried explicitly setting the arguments as you suggested, but that's not the problem. Vanilla importance sampling gives me estimates that are much closer to my MCMC. I think the SMC is working OK, as the ESS is high, so probably not driven by collapse.

```Julia
## Importance resampling

num_samples = 1000;
num_replicates = 1000;
β = Vector{Real}(undef, num_replicates);
i₀ = Vector{Int}(undef, num_replicates);
for i in 1:num_replicates
    (trace, _) = Gen.importance_resampling(bayes_sir_markov, (l,), observations, num_samples)
    β[i] = trace[:β]
    i₀[i] = trace[:i₀]
end;

mean(β),sqrt(var(β)) # (0.050589183232225526, 0.002358541039111068)
mean(i₀),sqrt(var(i₀)) # (16.366, 4.810112387583178)
```

As an aside, do I have to access the weights in the importance sample above in order to get the correct mean/variance? Any tips on how to extract the traced variables would also be helpful for debugging.

A grid-based approximation (esp. as I'm just trying to infer two parameters) would be useful too - any tips to get this set up would be great ([my vanilla particle filter](https://github.com/epirecipes/sir-julia/blob/master/markdown/markov_pfilter/markov_pfilter.md) does 2 x 1D sweeps, and you can see that the variance is much lower than my Gen.jl SMC setup.

**Reply by ztangent** [reply]
Interesting! Weird that SMC is so different from importance sampling. To get the weights for importance sampling, you can use `Gen.importance_sampling` instead of `Gen.importance_resampling`.

It's also possible that the `var(state::ParticleFilterState, addr)` convenience function isn't doing the right thing -- it's [currently returning a weighted empirical variance](https://github.com/probcomp/GenParticleFilters.jl/blob/f4fd6cf6cc4361a9aa56dd580a374f8f067658a2/src/statistics.jl#L41-L50), and when I implemented that, I didn't check whether this would return the same thing as drawing a lot of samples from the particle filter approximation (e.g. using `Gen.sample_unweighted_traces`) and computing the variance of those samples.

If you plot a histogram of the distribution over the inferred parameters for both importance sampling and SMC, do they look similar?

Will follow up with an example of computing a grid approximation.

**Reply by ztangent** [reply]
Okay here's how to do a grid-based approximation for a single-timestep version of your model:

```julia
@gen function bayes_sir_init(N::Int=1000, c::Float64=10.0, γ::Float64=0.25, δt::Float64=1.0)
    i₀ ~ uniform_discrete(1, 100)
    β ~ uniform(0.01, 0.1)
    S = N - i₀
    I = i₀
    ifrac = 1-exp(-β*c*I/N*δt)
    rfrac = 1-exp(-γ*δt)
    infection = {:y} ~ binom(S,ifrac)
    recovery = {:z} ~ binom(I,rfrac)
    S = S-infection
    I = I+infection-recovery
    return nothing
end
```

If we only observe `y`, then the latent variables we need to grid over are `i₀`, `β` and `z`. We can create a grid of choicemaps across this whole space using `GenParticleFilters.choiceproduct`, which returns an iterator over choicemaps. These are going to serve as our strata for stratified sampling:

```julia
strata = choiceproduct((:i₀, 1:100), (:β, 0.01:0.01:0.1), (:z, 0:100))
```

This will grid over `100 * 10 * 100 = 10^5` possible combinations of the latent variables (including some zero-probability combinations since `z` is supposed to be smaller than `i₀` --- you can be a bit smarter in how you grid if you're willing to write more specialized code that eliminates these impossible combinations.).  We can now initialize a particle collection that contains exactly one trace for each of these combinations by using the stratified version of `pf_initialize`:

```julia
obs_choices = choicemap((:y, 20))
n_particles = length(strata)
pf = pf_initialize(bayes_sir_init, (1000, 10.0, 0.25, 1.0), obs_choices, strata, n_particles)
```

Now we can look at the full marginal distributions over the latent variables we care about using e.g. `proportionmap`:

```julia
β_dist = proportionmap(pf, :β)
i₀_dist = proportionmap(pf, :i₀)
```

We can also compute the mean and variance as you've been doing, with the caveat that `var` might not be implemented in the way you expect.

To generalize this to a multi-timestep model, things get more complicated because now you have multiple `z` variables to grid over, and the problem explodes. You can probably afford to do this for a few timesteps just to double check that things are working. But if you want to get something that is closer-to-exact inference without requiring exponential amounts of memory, a scheme you could try is as follows:
- Initialize the particle filter by gridding over `choiceproduct((:i₀, 1:100), (:β, 0.01:0.01:0.1)`) at `t = 0`.
- With each new observation at step `t`, first use `pf_replicate!` to make `K` copies of every particle.
- Then use the stratified version of `pf_update!` to grid over all possible `K` new values of  `(:z, t)`, e.g. `choiceproduct(((:z, t), 1:100))` if `K= 100`.
- Use `pf_optimal_resize!` to resample the number of particles back down to the original number of particles (which should be 1000 in this case). Since this uses optimal resampling, the highest weight particles will all be kept around.
- Repeat until the last observation.

This procedure can be viewed as a low variance SMC algorithm that is similar to beam search with a beam-width of `K=100`, except that instead of always keeping the top `N` candidates, you use optimal resampling to select `N` candidates to keep, ensuring that the particle collection is always a properly-weighted approximation of the posterior. 

And of course, there are many variants of the above algorithm that trade off between variance and speed. For example, perhaps it'll be enough to use the non-stratified version of `pf_update!` in this case instead of gridding over all values of `(:z, t)`. In that case you can use a smaller value of `K`. If you make that change, you can also use `pf_residual_resize!` instead of `pf_optimal_resize!`, which keeps around multiple copies of high-weight particles, allowing them to have more descendants.

**Reply by sdwfrost** [reply]
Thanks; in order to try this out, I got rid of the process noise in my example (and just have some uncorrelated measurement error according to a Poisson) to reduce the need to simulate from `z`. In doing so, it seems that the SMC is just behaving really badly. With a seed fixed, I get OK results on the first try, and bad ones on the second. I just need to work harder on the SMC algorithm (which surprised me, as I got good results with IS and MCMC with MH pretty much out of the box). I really appreciate your help! For completeness, I'll work on the grid search (which should be feasible with the below implementation), and work up a better algorithm for the full Markov simulation. For completeness, it would be great to have a variational inference example - any starting points for that would also be appreciated.

```Julia
using OrdinaryDiffEq
using SciMLSensitivity
using Distributions
using Random
using StatsBase
using Gen
using GenDistributions
using GenParticleFilters
using Plots

Random.seed!(1234)

# Define the Gen model
@gen function bayes_sir_map(l::Int=40,
                               N::Float64=1000.0,
                               c::Float64=10.0,
                               γ::Float64=0.25,
                               δt::Float64=1.0)
    i₀ ~ uniform(1.0, 100.0)
    β ~ uniform(0.01, 0.1)
    S = N - i₀
    I = i₀
    for i in 1:l
        ifrac = 1-exp(-β*c*I/N*δt)
        rfrac = 1-exp(-γ*δt)
        infection = ifrac * S
        recovery = rfrac * I
        obs_infection = {(:y, i)} ~ poisson(infection)
        S = S-infection
        I = I+infection-recovery
    end
    nothing
end

# Set the true parameters and solve the model
p = Gen.choicemap()
p[:β] = 0.05
p[:i₀] = 10.0
fixed_args = (1000.0, 10.0, 0.25, 1.0)
l = 40
(sol, _) = Gen.generate(bayes_sir_map, (l,fixed_args...), p)

# Generate observations
Y = [sol[(:y, i)] for i=1:l]
observations = Gen.choicemap()
for (i, y) in enumerate(Y)
    observations[(:y, i)] = y
end

plot(1:l,Y)

## Importance

num_samples = 1000
num_replicates = 1000
β_is = Vector{Real}(undef, num_replicates)
i₀_is = Vector{Real}(undef, num_replicates)
for i in 1:num_replicates
    (trace, _) = Gen.importance_resampling(bayes_sir_map, (l,fixed_args...), observations, num_samples)
    β_is[i] = trace[:β]
    i₀_is[i] = trace[:i₀]
end

mean(β_is),sqrt(var(β_is)) # (0.04945130435442193, 0.00172879987091741)
mean(i₀_is),sqrt(var(i₀_is)) # (10.246874395245447, 1.9012491588244447)

histogram(β_is)
histogram(i₀_is)

## Metropolis-Hastings

const truncated_normal = DistributionsBacked((mu,std,lb,ub) -> Distributions.Truncated(Normal(mu, std), lb, ub),
                                             (true,true,false,false),
                                             true,
                                             Float64)


@gen function sir_proposal(current_trace)
    β ~ truncated_normal(current_trace[:β], 0.01, 0.0, Inf)
    i₀ ~ uniform(current_trace[:i₀]-1.0, current_trace[:i₀]+1.0)
end

n_iter = 110000
burnin =  10000
thin = 10
β_mh = Vector{Real}(undef, (n_iter - burnin) ÷ thin)
i₀_mh = Vector{Real}(undef, (n_iter - burnin) ÷ thin)

(tr,) = Gen.generate(bayes_sir_map, (l,fixed_args...), merge(observations, p))
for i in 1:n_iter
    (tr, _) = Gen.mh(tr, sir_proposal, ())
    if i <= burnin
        continue
    end
    if i % thin == 0
        β_mh[(i - burnin) ÷ thin] = tr[:β]
        i₀_mh[(i - burnin) ÷ thin] = tr[:i₀]
    end
end

# Posterior mean estimates
mean(β_mh), sqrt(var(β_mh)) # (0.04958329817997817, 0.0011092184431894826)
mean(i₀_mh), sqrt(var(i₀_mh)) # (9.937046546040595, 1.2154588575480654)

histogram(β_mh)
histogram(i₀_mh)

# Sequential Monte Carlo

function particle_filter(observations, n_particles, ess_thresh=0.5)
    # Initialize particle filter with first observation
    n_obs = length(observations)
    obs_choices = [choicemap((:y, t) => observations[t]) for t=1:n_obs]
    state = pf_initialize(bayes_sir_map, (1,), obs_choices[1], n_particles)
    # Iterate across timesteps
    for t=2:n_obs
        # Resample if the effective sample size is too low
        if effective_sample_size(state) < ess_thresh * n_particles
            # Perform residual resampling, pruning low-weight particles
            pf_resample!(state, :residual)
        end
        # Update filter state with new observation at timestep t
        new_args = (t, fixed_args...)
        argdiffs = (UnknownChange(), map(x -> NoChange(), new_args)...)
        pf_update!(state, new_args, argdiffs, obs_choices[t])
    end
    return state
end

n_particles = 10000
state = particle_filter(Y, n_particles)
effective_sample_size(state)
mean(state, :β), sqrt(var(state, :β)) # (0.05014782775663053, 0.0027900348439231)
mean(state, :i₀), sqrt(var(state, :i₀)) # (9.73524717317688, 6.312391801246316)



## Unweighted histogram

β_smc = getindex.(state.traces, :β)
i₀_smc = getindex.(state.traces, :i₀)

histogram(β_smc)
histogram(i₀_smc)

## Weighted histogram

w = get_norm_weights(state)
histogram(β_smc, weights=w)
histogram(i₀_smc, weights=w)

## Second try

state = particle_filter(Y, n_particles)
effective_sample_size(state)
mean(state, :β), sqrt(var(state, :β)) # (0.045352401808297804, 0.009407262234231674)
mean(state, :i₀), sqrt(var(state, :i₀)) # (30.33446829590008, 30.678623347235877)
```

**Reply by ztangent** [reply]
Happy to help! And interesting results, thanks! I wonder if the difference in performance is simply because both importance sampling and MCMC are using 10 to 100x more computation than the current SMC algorithm. If you ran 1000 replicates of SMC, each using 1000 particles, and then draw one particle from each replicate using `Gen.sample_unweighted_traces`, I would expect this to be better than just using `importance_resampling`.

For variational inference, the current support in Gen.jl is not as strong as we'd like, but there are some examples of black-box variational inference in the test cases: https://github.com/probcomp/Gen.jl/blob/master/test/inference/variational.jl

**Reply by sdwfrost** [reply]
Just using 100 replicates (due to time), with 1000 particles, I still get quite a few outliers. Can I ask why to use `Gen.sample_unweighted_traces`, rather than use the weights?

```Julia
# Iterative SMC

num_replicates = 100
num_particles = 1000
β_smc2 = Vector{Real}(undef, num_replicates)
i₀_smc2 = Vector{Real}(undef, num_replicates)
for i in 1:num_replicates
    state = particle_filter(Y, num_particles)
    trace = sample_unweighted_traces(state,1)[1]
    β_smc2[i] = trace[:β]
    i₀_smc2[i] = trace[:i₀]
end

mean(β_smc2),sqrt(var(β_smc2)) # (0.049592288660781, 0.007121180591284114)
mean(i₀_smc2),sqrt(var(i₀_smc2)) # (14.391547526358067, 17.99118514431295)

histogram(β_smc2)
histogram(i₀_smc2)
```

**Reply by ztangent** [reply]
Interesting that you're still getting outliers! I suspect that you are actually getting particle filter collapse a non-trivial amount of time, leading to the outliers. Effective sample size isn't enough to track whether that has happened, since after you perform a resampling move, the weights of the particles all go back to being equal, which means effective sample size goes back to the number of particles in the filter --- even if all of them are identical copies of each other. So what might be happening here is that on least some runs, the particle filter collapses on a value of i₀ that initially seems good after e.g. 5 observations, but turns out to be a bad explanation of the remaining 35 observations. Since the particle filter currently doesn't have any rejuvenation moves that perturb the value of i₀, there's no way to recover from this situation. 

A couple of ways to address would be to:
- Remove the resampling step altogether (this makes the particle filter equivalent to important sampling, except with slower runtime if you don't use the Unfold combinator)
- Resample less regularly --- e.g. check the resampling condition only every 5 or 10 timesteps, instead of every timestep.
- Add MCMC rejuvenation moves every few timesteps.

As for why to use `Gen.sample_unweighted_traces`, that's just to make a fair comparison with `Gen.importance_resampling`. `Gen.importance_resampling` first samples a collection of 1000 importance-weighted traces, then finally samples one trace from that collection according to its normalized importance weight. The particle filter also produces such a collection of 1000 importance-weighted traces. You can either use that weighted collection as your approximation to the posterior --- or, you can draw one particle from it (according to its normalized particle weight), giving you an unweighted sample that should look like it comes from something close to the true posterior. Doing this multiple times mimics what you've done with importance resampling.

**Reply by sdwfrost** [reply, marked as answer]
[Here](https://github.com/epirecipes/sir-julia/blob/master/markdown/markov_gen/markov_gen.md) is my example to date - I want to show packages in their best light, so suggestions to improve are much appreciated!
